File: HOT 100/1.go
/*
 * Created by Reverie
 * 2020.07.22
 * 21:51:52
 */

package main

import (
	"fmt"
	"sort"
)

//暴力扫描,时间复杂度O(n^2),空间复杂度O(1)
func twoSum1(nums []int, target int) []int {
	n := len(nums)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return []int{}
}

type myInt struct {
	idx int
	val int
}
type intSlice []myInt

func (ints intSlice) Len() int {
	return len(ints)
}
func (ints intSlice) Less(i, j int) bool {
	return ints[i].val < ints[j].val
}
func (ints intSlice) Swap(i, j int) {
	ints[i], ints[j] = ints[j], ints[i]
}

//排序+双指针,时间复杂度O(nlogn),空间复杂度O(n)
func twoSum2(nums []int, target int) []int {
	var ints intSlice
	for i, v := range nums {
		ints = append(ints, myInt{i, v})
	}
	sort.Sort(ints)
	i, j := 0, len(ints)-1
	for i < j {
		if ints[i].val+ints[j].val < target {
			i++
		} else if ints[i].val+ints[j].val > target {
			j--
		} else {
			return []int{ints[i].idx, ints[j].idx}
		}
	}
	return []int{}
}

//哈希表,时间复杂度O(n),空间复杂度O(n)
func twoSum3(nums []int, target int) []int {
	appear := make(map[int]int, len(nums))
	for i, v := range nums {
		if idx, ok := appear[target-v]; ok {
			return []int{idx, i}
		} else {
			appear[v] = i
		}
	}
	return []int{}
}

func main() {
	ans := twoSum3([]int{2, 7, 11, 15}, 9)
	fmt.Println(ans)
}

File: HOT 100/1.cpp
/*
 * Created by Reverie
 * 2020.07.22
 * 21:04:50
 */

#include <bits/stdc++.h>
using namespace std;

//暴力扫描,时间复杂度O(n^2),空间复杂度O(1)
class Solution1 {
  public:
    vector<int> twoSum(vector<int> &nums, int target) {
        vector<int> ans;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) {
                    ans.push_back(i);
                    ans.push_back(j);
                    return ans;
                }
            }
        }
        return ans;
    }
};

//排序+双指针,时间复杂度O(nlogn),空间复杂度O(n)
class Solution2 {
  public:
    vector<int> twoSum(vector<int> &nums, int target) {
        vector<int> ans;
        vector<int> temp = nums;
        sort(nums.begin(), nums.end());
        int i = 0, j = (int)nums.size() - 1;
        while (i < j) {
            if (nums[i] + nums[j] < target) {
                i++;
            } else if (nums[i] + nums[j] > target) {
                j--;
            } else {
                break;
            }
        }
        if (i < j) {
            for (int k = 0; k < nums.size(); k++) {
                if (temp[k] == nums[i]) {
                    ans.push_back(k);
                } else if (temp[k] == nums[j]) {
                    ans.push_back(k);
                }
            }
        }
        return ans;
    }
};

//哈希表,时间复杂度O(n),空间复杂度O(n)
class Solution3 {
  public:
    vector<int> twoSum(vector<int> &nums, int target) {
        vector<int> ans;
        unordered_map<int, int> appear;
        for (int i = 0; i < nums.size(); i++) {
            if (appear[target - nums[i]] == 0) {
                appear[nums[i]] = i + 1;
            } else {
                ans.push_back(i);
                ans.push_back(appear[target - nums[i]] - 1);
            }
        }
        return ans;
    }
};

// test
int main() {
    Solution3 s;
    vector<int> v = {2, 7, 11, 15};
    auto ans = s.twoSum(v, 9);
    if (ans.size() == 2) {
        cout << ans[0] << " " << ans[1] << endl;
    }
    return 0;
}